Parse opreturn before other outputs

sqrrm opened the pull request:
Since the effects of opreturn act on the BSQ outputs it makes more sense
to parse the opreturn before parsing the remaining outputs.

Put the lock time in the LOCKUP txOutput. It makes more logical sense to
keep it there.

sqrrm committed:
Parse opreturn before other outputs

Since the effects of opreturn act on the BSQ outputs it makes more sense
to parse the opreturn before parsing the remaining outputs.

Put the lock time in the LOCKUP txOutput. It makes more logical sense to
keep it there.

ManfredKarrer commented on src/main/java/bisq/core/dao/state/blockchain/Tx.java:
```diff
@@ -138,14 +138,12 @@ public TxOutput getLastTxOutput() {
-     * OpReturn output might contain the lockTime in case of a LockTx. It has to be the last output.
-     * We store technically the lockTime there as is is stored in the OpReturn data but conceptually we want to provide
-     * it from the transaction.
+     * The locktime is stored in the LOCKUP txOutput, which is the first txOutput.
```
Hm... I am not sure if the first output is a good place here. Technically its the opReturn output, conceptually its the tx (thats why we use a delegate there).

sqrrm replied:
The reason I moved it to the LOCKUP txoutput is that elsewhere it's expected to reside there. It's also a logical place since that's the txoutput that has a restriction on it.

ManfredKarrer commented on src/main/java/bisq/core/dao/node/parser/TxParser.java:
```diff
@@ -154,6 +158,9 @@ public TxParser(PeriodService periodService,
             } else {
+                // TODO(SQ): The transaction has already been parsed here and the individual txouputs are considered
+                // spendable or otherwise correct. Perhaps this check should be done earlier.
+
```
I think you are correct. Should be probably done earlier

ManfredKarrer commented on src/main/java/bisq/core/dao/node/parser/TxParser.java:
```diff
@@ -115,11 +115,15 @@ public TxParser(PeriodService periodService,
-            txOutputParser.processOpReturnCandidate(outputs.get(lastIndex));
+            int lastNonOpReturnIndex = lastIndex;
+            if (txOutputParser.processOpReturnCandidate(outputs.get(lastIndex))) {
+                txOutputParser.processTxOutput(true, outputs.get(lastIndex), lastIndex);
```
I am not 100% if that can cause problems. We can discuss tomorrow.

sqrrm replied:
From what I see there is only the lockup opreturn processing that really does anything here, and that was the one I wanted to fix. Would be good with more eyes to verify that of course.

ManfredKarrer commented on src/main/java/bisq/core/dao/node/parser/TxOutputParser.java:
```diff
@@ -81,8 +82,9 @@ public void processGenesisTxOutput(TempTx genesisTx) {
     }
 
-    void processOpReturnCandidate(TempTxOutput txOutput) {
+    boolean processOpReturnCandidate(TempTxOutput txOutput) {
```
If we return a boolean we should rename to make that more clear (e.g. isOpReturnCandidate)

ManfredKarrer commented on src/main/java/bisq/core/dao/node/parser/TxOutputParser.java:
```diff
@@ -164,6 +166,7 @@ private void handleBsqOutput(TempTxOutput txOutput, int index, long txOutputValu
             optionalVoteRevealUnlockStakeOutput = Optional.of(txOutput);
         } else if (isFirstOutput && opReturnTypeCandidate == OpReturnType.LOCKUP) {
             bsqOutput = TxOutputType.LOCKUP;
+            txOutput.setLockTime(lockTime);
```
Why put it in first output? Technically its in opReturn, conceptually in Tx.

sqrrm committed:
Rename processOpReturnCandidate -> isOpReturnCandidate

sqrrm committed:
Extract opreturn post processing functions

sqrrm committed:
Process opReturn txOutput before remaining outputs

sqrrm committed:
Check for multiple opReturns before processing txOutputs

sqrrm committed:
Wait to commit BSQ txOutputs to bsqStateService until after parsing

If a transaction is considered invalid it should consider all the
txOutputs to be invalid. To avoid counting a txOutput as valid it's
better to commit to stateservice only after checking the validity of the
transaction.

sqrrm committed:
Merge remote-tracking branch 'origin/master' into rearrange-bsq-parsing

sqrrm commented:
I did some more refactoring, the main thing being that the txOutputs are only commited after the tx type is determined. That way we avoid inconsistency of adding some outputs to bsq stateservice and later finding out the tx is invalid.
